### PR TITLE
Introduce Collectors to (dramatically) speed up consistency checks

### PR DESCRIPTION
**Pull request by little-bobby-tables**
Running `mix credo` on the master branch takes 10.5 seconds on my machine; `mix credo --ignore consistency` drops that to 4.6 seconds. With this branch, `mix credo` takes 4.7 seconds.

The slowest checks are `TabsOrSpaces` and `LineEndings`. They attach lots of meta information (filename, line_no) to _every_ match, only to discard it later.

This PR introduces a `Collector` module that provides general functions for reducing the input to a map of `%{match => frequency}`, e.g. `%{with_spaces: 300, without_spaces: 2}`, and then obtaining the required meta information on a per-file basis, for files that are known to have issues.

I think most of the consistency checks could benefit from this structure as it reduces the amount of code repetition and indirection, although there are exceptions like `MultiAliasImportRequireUse` which do not fit in.

I understand it if you don't fully agree with the implementation — I'm open to discussion and  making changes deemed necessary.

### CODE REVIEW

File: lib/credo/check/consistency/line_endings.ex [hunk 1]
```diff
@@ -8,31 +8,25 @@ defmodule Credo.Check.Consistency.LineEndings do
...
-  defp issue_for(_issue_meta, [], _expected_prop, _picked_count, _total_count), do: nil
-  defp issue_for(issue_meta, actual_prop, expected_prop, _picked_count, _total_count) do
-    actual_prop = PropertyValue.get(actual_prop)
-
-    format_issue issue_meta,
-      message: "File is using #{actual_prop} line endings while most of the files use #{expected_prop} line endings."
+  defp issues_for(expected, {[actual], source_file, params}) do
+    source_file
+    |> IssueMeta.for(params)
+    |> format_issue(message: "File is using #{actual} line endings\
```
Comment (rrrene) [inline lib/credo/check/consistency/line_endings.ex]: Please do not break the binary. Just let it be more than 80 chars, it's no big deal.

File: lib/credo/check/consistency/collector.ex [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      @spec insert_issue(Issue.t, Credo.Execution.t) :: term
+      def insert_issue(%Issue{filename: filename} = issue, exec) do
+        Credo.Execution.Issues.append(
+          exec, %SourceFile{filename: filename}, issue)
+      end
+    end
+  end
+
+  def issues(source_files, collector, params, issue_formatter) do
+    frequencies_per_file = Enum.map(source_files, fn(file) ->
```
Comment (rrrene) [inline lib/credo/check/consistency/collector.ex]: Please format this multi-line assignment like the next one (break after `=`).

File: lib/credo/check/consistency/collector.ex [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      @spec find_issues(
+        [SourceFile.t], Keyword.t, Collector.issue_formatter) :: [Issue.t]
+      def find_issues(source_files, params, issue_formatter)
+        when is_list(source_files) and is_function(issue_formatter) do
+          Collector.issues(source_files, __MODULE__, params, issue_formatter)
+      end
+
+      @spec insert_issue(Issue.t, Credo.Execution.t) :: term
+      def insert_issue(%Issue{filename: filename} = issue, exec) do
+        Credo.Execution.Issues.append(
```
Comment (rrrene) [inline lib/credo/check/consistency/collector.ex]: Please alias the `Issues` module.

File: lib/credo/check/consistency/collector.ex [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      @behaviour Credo.Check.Consistency.Collector
+
+      alias Credo.Issue
+      alias Credo.SourceFile
+      alias Credo.Check.Consistency.Collector
+
+      @spec find_issues(
+        [SourceFile.t], Keyword.t, Collector.issue_formatter) :: [Issue.t]
+      def find_issues(source_files, params, issue_formatter)
+        when is_list(source_files) and is_function(issue_formatter) do
```
Comment (rrrene) [inline lib/credo/check/consistency/collector.ex]: Please put the `when` guard in the previous line.

File: lib/credo/check/consistency/collector.ex [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+          Collector.issues(source_files, __MODULE__, params, issue_formatter)
```
Comment (rrrene) [inline lib/credo/check/consistency/collector.ex]: Please fix the indentation here.

File: lib/credo/check/consistency/collector.ex [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    frequencies_per_file
+    |> issues_per_file(most_frequent, params)
+    |> Enum.flat_map(&issue_formatter.(most_frequent, &1))
+  end
+
+  defp issues_per_file(frequencies_per_file, most_frequent, params) do
+    Enum.reduce(frequencies_per_file, [], fn({file, frequencies}, with_issues) ->
+      invalid_values = Map.keys(frequencies) -- [most_frequent]
+      if invalid_values != [] do
```
Comment (rrrene) [inline lib/credo/check/consistency/collector.ex]: Please insert a blank line above this one.

File: lib/credo/check/consistency/collector.ex [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        [{invalid_values, file, params} | with_issues]
+      else
+        with_issues
+      end
+    end)
+  end
+
+  defp total_frequencies(frequencies_per_file) do
+    Enum.reduce(frequencies_per_file, %{}, fn({_, frequencies}, stats) ->
+      Map.merge(stats, frequencies, fn (_k, f1, f2) -> f1 + f2 end)
```
Comment (rrrene) [inline lib/credo/check/consistency/collector.ex]: Please remove the space after `fn`.

File: lib/credo/check/consistency/exception_names/prefix_and_suffix_collector.ex [hunk 8]
```diff
@@ -3,37 +3,35 @@ defmodule Credo.Check.Consistency.ExceptionNames.PrefixAndSuffixCollector do
...
+  defp property_value_for_exception_name(name, {_, meta, _}) do
     name_list = name |> Name.last |> Name.split_pascal_case
     prefix = List.first(name_list)
     suffix = List.last(name_list)
 
     [
-      PropertyValue.for({prefix, :prefix}, filename),
-      PropertyValue.for({suffix, :suffix}, filename)
+      PropertyValue.for({prefix, :prefix}, [line_no: meta[:line]]),
+      PropertyValue.for({suffix, :suffix}, [line_no: meta[:line]])
```
Comment (rrrene) [inline lib/credo/check/consistency/exception_names/prefix_and_suffix_collector.ex]: Let's remove the square brackets in the above 2 lines (just for consistency).

File: lib/credo/check/consistency/tabs_or_spaces/collector.ex [hunk 9]
```diff
@@ -0,0 +1,24 @@
+defmodule Credo.Check.Consistency.TabsOrSpaces.Collector do
+  use Credo.Check.Consistency.Collector
+
+  def collect_values(source_file, _params) do
+    source_file
+    |> SourceFile.lines
+    |> Enum.reduce(%{}, fn(line, stats) ->
+        match = indented_with(line)
+        if match, do: Map.update(stats, match, 1, &(&1 + 1)), else: stats
```
Comment (rrrene) [inline lib/credo/check/consistency/tabs_or_spaces/collector.ex]: Please put a blank line before this one.

**Comment by little-bobby-tables** [comment]
Oops, sorry for that :smiley: I wasn't sure how strict you feel about the 80 columns rule.

File: lib/credo/check/consistency/exception_names/prefix_and_suffix_collector.ex [hunk 10]
```diff
@@ -3,37 +3,35 @@ defmodule Credo.Check.Consistency.ExceptionNames.PrefixAndSuffixCollector do
 
   alias Credo.Code.Module
   alias Credo.Code.Name
-  alias Credo.CLI.Filename
 
   def property_value, do: nil
 
-  def property_value_for(%SourceFile{filename: filename} = source_file, _params) do
-    Credo.Code.prewalk(source_file, &traverse(&1, &2, filename))
+  def property_value_for(source_file, _params) do
```
Comment (rrrene) [inline lib/credo/check/consistency/exception_names/prefix_and_suffix_collector.ex]: Q: Why are we touching all the "code pattern" files of the otherwise untouched consistency checks?

File: lib/credo/check/consistency/collector.ex [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+  alias Credo.Issue
+  alias Credo.SourceFile
+
+  @doc """
+  `collect_values` returns a map of matches (atoms) and their frequencies
+  (e.g. %{with_space: 50, without_space: 40}) for a given file.
+  """
+  @callback collect_values(source_file :: SourceFile.t, params :: Keyword.t)
+    :: %{atom => non_neg_integer} | %{}
```
Comment (little-bobby-tables) [inline lib/credo/check/consistency/collector.ex]: Consistency checks that I was interested in all used atoms, so I went with that. A quick microbenchmark doesn't show substantial difference between atoms, binaries, and non-empty lists used as keys. Do you think this requirement should be relaxed?

File: lib/credo/check/consistency/collector.ex [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+  @doc """
+  `find_locations` returns metadata for each occurrence of a match
+  in a given file. It is called only when the file is known to contain
+  the match.
+  """
+  @callback find_locations(match :: atom, source_file :: SourceFile.t)
```
Comment (little-bobby-tables) [inline lib/credo/check/consistency/collector.ex]: The return value may include column number and trigger (as in case of `SpaceAroundOperators`). I agree that for the two checks that I did touch it alone it doesn't make sense to keep it this way, but since I'll be refactoring other modules, it's going to be used in more places.

File: lib/credo/check/consistency/exception_names/prefix_and_suffix_collector.ex (hunk 10 above)
Comment (little-bobby-tables) [inline lib/credo/check/consistency/exception_names/prefix_and_suffix_collector.ex]: Including filenames in property values seems redundant to me, as they're already matched against the corresponding source files in `Credo.Check.Consistency.Helper.create_property_tuples`.

This may be out of scope of this PR though, do you want me to revert these changes?

File: lib/credo/check/consistency/collector.ex (hunk 12 above)
Comment (rrrene) [inline lib/credo/check/consistency/collector.ex]: :+1: Thx for the clarification!

File: lib/credo/check/consistency/collector.ex (hunk 11 above)
Comment (rrrene) [inline lib/credo/check/consistency/collector.ex]: I'd like to see this "Collector" concept in the other checks as well. But to collect e.g. the prefixes and suffixes for the `ExceptionNames` check, it would be nicer to have Erlang terms instead of atoms (if it is okay performance wise).

File: lib/credo/check/consistency/exception_names/prefix_and_suffix_collector.ex (hunk 10 above)
Comment (rrrene) [inline lib/credo/check/consistency/exception_names/prefix_and_suffix_collector.ex]: No no, I just wanted to confirm that I understood it correctly 😉

**Comment by rrrene** [comment]
Great. Do you have time to adapt this for the other consistency checks?

For performance "benchmarking" I try to not run Credo on itself to reduce bias. For "real" projects, you can run it on Plug (which takes around 10 secs on your branch) or Elixir itself (use `lib/elixir` to analyse just a subset of Elixir's source, which takes roughly 200 seconds or analyse the full source for a longer analysis, taking up to half an hour on my slower laptop).

**Comment by little-bobby-tables** [comment]
Yes, I'll try to refactor other consistency checks as well. `TabsOrSpaces` and `LineEndings` were just low-hanging fruit, they were the easiest to change and it yielded the biggest performance improvement.

**Comment by little-bobby-tables** [comment]
@rrrene, I went through the remaining consistency checks and implemented collectors for them. How do you feel about removing `Consistency.Helper`?

**Comment by rrrene** [comment]
This looks very good.

I think we should improve the documentation a bit, to show people how they are supposed to use the "Collector pattern" in their own checks. Credo allows you to create your own checks and documenting the relevant parts of the code for this is an important task to reach Credo v1.0 😄

**Comment by little-bobby-tables** [comment]
Oh, I see. Do you think I should explain the structure in `collector.ex` or perhaps create something like `.template.check.ex`? The latter might be a little more convenient in my opinion since there _is_ some boilerplate involved.

**Comment by rrrene** [comment]
I think we should do our best to include this in the documentation. 

`Plug` does a great job of explaining itself in its docs and we should try to do that as well.

Great examples are [the Plug behaviour](https://github.com/elixir-lang/plug/blob/v1.3.5/lib/plug.ex) and [Plug.Router](https://github.com/elixir-lang/plug/blob/v1.3.5/lib/plug/router.ex).

**Comment by rrrene** [comment]
@little-bobby-tables this is looking really good! I will review this more closely ASAP :+1:

**Comment by little-bobby-tables** [comment]
Okay, thanks! Let me know if anything else needs to be done.

**Comment by rrrene** [comment]
@little-bobby-tables Hi, super sorry this took so long. I had to get married, find a house and defend my PhD thesis.

But: you contributions are now part of Credo `v0.8.2`, which I published last night. Thx again for your work! :+1:

**Comment by little-bobby-tables** [comment]
Thanks, @rrene! My congratulations on your success! :smiley: Real life comes first, and it's perfectly understandable.
